### ISSUE TITLE
lintr appears to be breaking the package build

### ISSUE THREAD

**Issue by swager** [grf-labs/grf#471]
Below is what my session looked like. R abruptly exits after running `lintr`.

```
GSB-C02TP1BZGTDX:r-package swager$ R

R version 3.4.0 (2017-04-21) -- "You Stupid Darkness"
Copyright (C) 2017 The R Foundation for Statistical Computing
Platform: x86_64-apple-darwin15.6.0 (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

  Natural language support but running in an English locale

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

> source("build_package.R")

Attaching package: ‘lintr’

The following object is masked from ‘package:devtools’:

    lint

...........................................
Save workspace image? [y/n/c]: n
Warning messages:
1: package ‘Rcpp’ was built under R version 3.4.4 
2: package ‘devtools’ was built under R version 3.4.4 
3: package ‘lintr’ was built under R version 3.4.4 
GSB-C02TP1BZGTDX:r-package swager$ 
```

**Comment by erikcs** [comment]
Yes, [this line](https://github.com/grf-labs/grf/blob/master/r-package/build_package.R#L30) quits R in case there is a lint error, please run it as  `Rscript build_package.R` instead.

**Comment by swager** [comment]
Ah, got it. My take is that the linter really shouldn't be halting the build ever -- rather, it should just print a clear warning with the lint error and then proceed.

The worst case scenario is if someone is developing under time pressure, and then can't get the build to go through because of a linter error. Linter errors should eventually be fixed (e.g., before pushing to master), but we don't need to require the linter to pass at all times while prototyping new features. The linter should be a friend, not a foe.

**Comment by erikcs** [comment]
Yes, it can get annoying. It was put in place to fail the PR early (it needs to quit in order for Travis to register the lintr error as a failure). Should I move the quit statement to the last line?

(By the way, another quick alternative to the build_package.R is having the grf.Rproj open and running Build-> Load all, or Clean and Rebuild, then Build->Test Package)

**Comment by swager** [comment]
Relatedly, I was just working on #472, and travis failed because of a linter error (without providing a particularly helpful error message).

![image](https://user-images.githubusercontent.com/1275702/62484753-b2c4b700-b76f-11e9-96ca-46d2bbd3f607.png)

I understand that a good linter can really speed up the development process, but from my experience so far I'm not yet convinced the lintr package is up to the task. How do the rest of you feel?

I think a happy middle ground could be if lintr were only giving warnings (but not making builds fail).

**Comment by jtibshirani** [comment]
Could we refactor the build script to group the build steps into a method, then simply return from the method early if lintr fails instead of a hard session termination? I think that error is due to the use of `=` instead of `<-` within a test: https://github.com/jimhester/lintr/issues/391. The message is very cryptic and may be a regression in lintr.

> I think a happy middle ground could be if lintr were only giving warnings (but not making builds fail).

I would be also be okay with this, we could just print out the warnings as part of the build process. When reviewing PRs, it will be easy to point out formatting errors by having the author look at the build printout (or asking them to address the warnings they see locally).

**Comment by erikcs** [comment]
I can open a PR a bit later. lintr is definitely nowhere near for example flake8 for python, but was the only alternative (over nothing).

**Comment by halflearned** [comment]
I wouldn’t shed any tears if the linter were removed altogether. 

My opinion is that a linter is useful if the language has a reasonably strong consensus for what constitutes "correct" style, so it's not difficult to get most things right from the first time we sit down to write code. I find myself disagreeing with the current linter a lot, and I often don't see the error messages until I actually build the package (or worse, when I've already pushed the PR).
